source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst

```diff
@@ -355,29 +402,93 @@ Before a connection can be returned outside of the pool, it must be “set up”
 Closing a Connection (Internal Implementation)
 ----------------------------------------------
 
-When a Connection is closed, it MUST first be marked as closed, removing it from being counted as available, in use, or setup. Once the connection is marked as closed, the Connection can perform whatever teardown is necessary. The Driver SHOULD perform the teardown in a non-blocking manner.
+When a `Connection <#connection>`_ is closed, it MUST first be marked as "closed",
+removing it from being counted as "available" or "in use". One that is
+complete, the `Connection <#connection>`_ can perform whatever teardown is
+necessary to close its underlying socket. The Driver SHOULD perform this
+teardown in a non-blocking manner, such as via the use of a background
+thread or async I/O.
 
 .. code::
 
-    mark connection as closed
+    original state = connection state
+    set connection state to "closed"
+    decrement total connection count
+    if original state is "available":
+      decrement available connection count
     emit ConnectionClosedEvent
 
-    # The following can happen at a later time
-    connection.close()
+    # The following can happen at a later time (i.e. in background
+    # thread) or via non-blocking I/O.
+    connection.socket.close()
+
+Marking a Connection as Available (Internal Implementation)
+-----------------------------------------------------------
+
+A `Connection <#connection>`_ is "available" if it is able to be checked out. A
+`Connection <#connection>`_ MUST NOT be marked as "available" until it has been
+established. The pool MUST keep track of the number of currently
+available `Connections <#connection>`_.
+
+.. code::
+
+   increment available connection count
+   set connection state to "available"
+   add connection to availableConnections
+
+
+Populating the Pool with a Connection (Internal Implementation)
+---------------------------------------------------------------
+
+"Populating" the pool involves creating and establishing a `Connection
+<#connection>`_ that will not immediately be checked out and will instead be
+marked as "available" and kept in the pool. This type of `Connection
+<#connection>`_ is created to ensure minPoolSize and is distinct from from the
+type that is created, established, and immediately checked out in checkOut. See
```

Comment:
Conceptually I consider all connections to be of the same type and go through the same exact lifecycle. Hence it is weird to read "this type of connection is created...". I agree that the processes by which a Connection may be created can vary based on whether it is created in main thread or in a background thread.

Again I would simply drop this sentence, I am not seeing the value it adds to the spec.

Reply:
Yeah, I don't think classifying them according to type is the best way to word this. I was attempting to classify the context of the creation rather than the type of the Connection itself, since as you point out, there aren't different types of Connections. The two creation contexts I'm trying to disambiguate are "on-demand" (e.g. checkOut is called and pool has no available connections) and "preemptive" (e.g. ensuring minPoolSize, which is done during population). 

To that end, I've struck the sentence about different types of Connections and rewritten the section to focus more on the preemptive nature of these Connection creations. I hope that clarifies this section a little more.

Reply:
:+1: The revised wording is great.